File: packages/postpin-go/rates.go
package postpin

import (
	"context"
	"net/http"
)

// RatesService calculates shipping rates.
type RatesService struct{ client *Client }

// Calculate computes the shipping rate between two Indian pincodes.
func (s *RatesService) Calculate(ctx context.Context, params RateParams) (*RateResult, error) {
	body := rateBody{
		Origin:        params.Origin,
		Destination:   params.Destination,
		Weight:        params.Weight,
		Length:        params.Length,
		Width:         params.Width,
		Height:        params.Height,
		Service:       params.Service,
		COD:           params.COD,
		DeclaredValue: params.DeclaredValue,
	}
	data, err := s.client.do(ctx, http.MethodPost, "/rates/calculate", nil, body, params.IdempotencyKey)
	if err != nil {
		return nil, err
	}
	var out RateResult
	if err := decode(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: packages/postpin-node/src/resources/rates.ts
import type { HttpClient } from "../http.js";
import type { RateCalculateParams, RateResult, RequestOptions } from "../types.js";

/** Shipping-rate calculation — the core, billable endpoint. */
export class Rates {
  constructor(private readonly http: HttpClient) {}

  /**
   * Calculate a shipping rate between two pincodes.
   *
   * @example
   * const rate = await postpin.rates.calculate({
   *   origin: "400001",
   *   destination: "110001",
   *   weight: 1200,
   *   service: "express",
   * });
   * console.log(rate.total, rate.zoneLabel);
   */
  async calculate(params: RateCalculateParams, options: RequestOptions = {}): Promise<RateResult> {
    const { data } = await this.http.request<RateResult>({
      method: "POST",
      path: "/rates/calculate",
      body: {
        origin: params.origin,
        destination: params.destination,
        weight: params.weight,
        length: params.length,
        width: params.width,
        height: params.height,
        service: params.service,
        cod: params.cod,
        declared_value: params.declaredValue,
      },
      idempotencyKey: options.idempotencyKey,
      signal: options.signal,
    });
    return data;
  }
}
